### PR TITLE
#7: REST API Client

### PR DESCRIPTION
**Pull request by BeeMargarida**
| - | - |
| --- | --- |
| Issue | Closes https://github.com/ripe-tech/ripe-registry/issues/7|
| Dependencies | -- |
| Decisions | APIs for the routes in [ripe-registry](https://github.com/ripe-tech/ripe-registry) and respective type declarations and documentation. Unit tests for the PersonAPI, which is the only one not using mocked data. |
| Animated GIF | Screenshot 2020-08-03 at 16 44 57Screenshot 2020-08-03 at 16 45 05|

### CODE REVIEW

File: js/client.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            return client;
+        }
+
+        /**
+         * Updates the client with the provided information.
+         *
+         * @memberof ClientAPI
+         * @param {String} name The name of the client.
+         * @param {Object} body An object that contains information about a client, such as:
+         * - 'name' - The name of the client.
```
Comment (NFSS10) [inline js/client.js]: I'm not sure `name` is allowed in `body`

File: js/client.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+ */
+export const ClientAPI = superclass =>
+    class extends superclass {
+        /**
+         * Returns all clients in the registry.
+         *
+         * @memberof ClientAPI
+         * @returns {Promise} The client result list.
+         */
+        async getClients() {
```
Comment (NFSS10) [inline js/client.js]: Shouldn't it be `listClients()` ?

Same question for the other of endpoints that list stuff

File: js/person.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        }
+
+        /**
+         * Updates the person with the provided information.
+         *
+         * @memberof PersonAPI
+         * @param {String} email The e-mail of the person.
+         * @param {Object} body An object that contains information about a person, such as:
+         * - 'name' - The name of the person.
+         * - 'email' - The e-mail of the person.
```
Comment (NFSS10) [inline js/person.js]: I'm not sure if `email` is allowed in `body`

File: js/client.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        }
+
+        /**
+         * Returns the projects of the provided client.
+         *
+         * @memberof ClientAPI
+         * @param {String} email name The name of the client.
+         * @returns {Promise} The client's projects result list.
+         */
+        async getClientProjects(name) {
```
Comment (NFSS10) [inline js/client.js]: Shouldn't it be `listClientProjects(name)` ?

File: js/person.js [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        }
+
+        /**
+         * Returns the projects of the provided person.
+         *
+         * @memberof PersonAPI
+         * @param {String} email The e-mail of the person.
+         * @returns {Promise} The person's projects result list.
+         */
+        async getPersonProjects(email) {
```
Comment (NFSS10) [inline js/person.js]: Shouldn't it be `listPersonProjects(email)` ?

File: js/person.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        }
+
+        /**
+         * Returns the clients of the provided person.
+         *
+         * @memberof PersonAPI
+         * @param {String} email The e-mail of the person.
+         * @returns {Promise} The person's clients result list.
+         */
+        async getPersonClients(email) {
```
Comment (NFSS10) [inline js/person.js]: Shouldn't it be `listPersonClients(email)` ?

File: js/project.js [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+            return project;
+        }
+
+        /**
+         * Updates the project with the provided information.
+         *
+         * @memberof ProjectAPI
+         * @param {String} name The name of the project.
+         * @param {Object} body An object that contains the updated information about the project, such as:
+         * - 'name' - The unique project's name, used to identify and differentiate from the other projects.
```
Comment (NFSS10) [inline js/project.js]: I'm not sure if `name` is allowed in `body`

File: js/project.js [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+        }
+
+        /**
+         * Returns the iterations of the provided project.
+         *
+         * @memberof ProjectAPI
+         * @param {String} name The name of the project.
+         * @returns {Promise} The project's iterations result list.
+         */
+        async getProjectIterations(name) {
```
Comment (NFSS10) [inline js/project.js]: Shouldn't it be `listProjectIterations(name)`

File: test/api/base.js [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        assert.strictEqual(person.name, "Jason A. Barton");
+        assert.strictEqual(person.email, "jasonA@email.com");
+        assert.strictEqual(person.company, "Star Bright Investment Group");
+        assert.strictEqual(person.position, "Engineering technician");
+        assert.strictEqual(person.phone, "682-503-4335");
+
+        await api.deletePerson("jasonA@email.com");
+    });
+
+    it("should be able to list the existing person", async () => {
```
Comment (NFSS10) [inline test/api/base.js]: Maybe change "list" to "get"

File: test/api/base.js [hunk 10]
```diff
@@ -0,0 +1,82 @@
+const assert = require("assert");
+const ripe = require("../..");
+
+describe("PersonAPI", function() {
+    it("should be able to construct a new person", async () => {
+        const api = new ripe.API();
```
Comment (NFSS10) [inline test/api/base.js:6]: As the API is always the same, maybe move `const api = new ripe.API();` to line 5.
After`describe("PersonAPI", function() {` and before `it("should be able to construct a new person", async () => {`

File: types/client.d.ts [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+export interface Client {
+    readonly id: string;
+    readonly name: string;
+    readonly website: string;
+    readonly accountManager: Person;
+    readonly clientRepresentative: Person;
+}
+
+export declare class ClientAPI {
+    getClients(): Array<Client>;
```
Comment (NFSS10) [inline types/client.d.ts]: If you change `getClients` to `listClients`, don't forget to update the remaining interfaces

File: types/client.d.ts [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+    getClient(name: number): Client;
```
Comment (NFSS10) [inline types/client.d.ts]: `getClient(name: number): Client;` should be `getClient(name: string): Client;` right?

File: types/person.d.ts [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+    readonly phone: string;
+    readonly platformeId: string;
+}
+
+export declare class PersonAPI {
+    getPersons(): Array<Person>
+    getPerson(email: string): Person;
+    createPerson(body: object): Person;
+    updatePerson(email: string, body: object): Person;
+    deletePerson(email:string): Person;
```
Comment (NFSS10) [inline types/person.d.ts]: Missing space `email:string` -> `email: string`

**Review by NFSS10 — CHANGES_REQUESTED** [review]

**Review by NFSS10 — APPROVED** [review]

File: js/client.js [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+    class extends superclass {
+        /**
+         * Returns all clients in the registry.
+         *
+         * @memberof ClientAPI
+         * @returns {Promise} The client result list.
+         */
+        async listClients() {
+            const url = this.baseUrl + "clients";
+            console.log(url);
```
Comment (BeeMargarida) [inline js/client.js]: oops

**Review by gcandal — APPROVED** [review]

**Review by gcandal — APPROVED** [review]

File: types/project.d.ts [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+    readonly name: string;
+    readonly description: string;
+    readonly client: Client;
+    readonly manager: Person;
+}
+
+export declare class ProjectAPI {
+    listProjects(): Array<Project>
+    getProject(name: string): Project;
+    createProject(body: object): Project;
```
Comment (joamag) [inline types/project.d.ts]: use `Project` type  for the input (make id optional)

File: types/person.d.ts [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+    readonly company: string;
+    readonly position: string;
+    readonly phone: string;
+    readonly platformeId: string;
+}
+
+export declare class PersonAPI {
+    listPersons(): Array<Person>
+    getPerson(email: string): Person;
+    createPerson(body: object): Person;
```
Comment (joamag) [inline types/person.d.ts]: Use use `Person` type for the input (make id optional)

File: js/client.js [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+            return client;
+        }
+
+        /**
+         * Updates the client with the provided information.
+         *
+         * @memberof ClientAPI
+         * @param {String} name The name of the client.
+         * @param {Object} body An object that contains information about a client, such as:
+         * - 'website' - The client's website URL.
```
Comment (joamag) [inline js/client.js]: No need to comment on the fields of the payload using the type description instead.
Rename this to `payload` instead of boyd

**Review by joamag — CHANGES_REQUESTED** [review]

**Review by gcandal — APPROVED** [review]

File: types/client.d.ts [hunk 18]
```diff
@@ -0,0 +1,19 @@
+import { Person } from "./person";
+import { Project } from "./project";
+
+export interface Client {
```
Comment (joamag) [inline types/client.d.ts:4]: Create the `ClientPatch` interface that has all of the fields as optional.
Client the `ClientCreate` with the definition of mandatory and non-mandatory fields.

**Review by joamag — CHANGES_REQUESTED** [review]

**Review by gcandal — APPROVED** [review]

**Review by joamag — APPROVED** [review]

**Review by joamag — APPROVED** [review]

**Comment by joamag** [comment]
@gcandal let's try to have CI/CD on the esdocs generation using vercel

**Pull request merged**